File: openless-all/app/src/lib/ipc/cloud-sync.ts
import { invokeOrMock } from './shared';

export interface CloudSyncStatus {
  schemaVersion: number;
  revision: number;
  updatedAt: string | null;
  hasSnapshot: boolean;
  counts: { dictionary: number; corrections: number; stylePacks: number };
}

export interface CloudSyncUiPreferences {
  locale?: string;
  fontScale?: string;
}

export interface CloudSyncRestoreResult {
  status: CloudSyncStatus;
  uiPreferences: CloudSyncUiPreferences;
}

// A browser preview has no native stores or authenticated cloud account.
function unavailable(): never {
  throw new Error('cloud_sync_unavailable');
}

export function cloudSyncStatus(): Promise<CloudSyncStatus> {
  return invokeOrMock('cloud_sync_status', undefined, unavailable);
}

export function cloudSyncUpload(
  baseRevision: number,
  uiPreferences: CloudSyncUiPreferences,
): Promise<CloudSyncStatus> {
  return invokeOrMock('cloud_sync_upload', { baseRevision, uiPreferences }, unavailable);
}

export function cloudSyncRestore(): Promise<CloudSyncRestoreResult> {
  return invokeOrMock('cloud_sync_restore', undefined, unavailable);
}

export function cloudSyncDelete(baseRevision: number): Promise<CloudSyncStatus> {
  return invokeOrMock('cloud_sync_delete', { baseRevision }, unavailable);
}

File: openless-all/app/src-tauri/src/commands/cloud_sync.rs
//! Cloud operations delegate to Core so browser inputs never contain provider or GitHub credentials.

use super::CoreState;
use openless_core::{
    BackendError, CloudSyncRestoreResult, CloudSyncStatus, CloudSyncUiPreferences,
};

#[tauri::command]
pub async fn cloud_sync_status(core: CoreState<'_>) -> Result<CloudSyncStatus, BackendError> {
    core.cloud_sync_status().await
}

#[tauri::command]
pub async fn cloud_sync_upload(
    core: CoreState<'_>,
    base_revision: u64,
    ui_preferences: CloudSyncUiPreferences,
) -> Result<CloudSyncStatus, BackendError> {
    core.cloud_sync_upload(base_revision, ui_preferences).await
}

#[tauri::command]
pub async fn cloud_sync_restore(
    core: CoreState<'_>,
) -> Result<CloudSyncRestoreResult, BackendError> {
    core.cloud_sync_restore().await
}

#[tauri::command]
pub async fn cloud_sync_delete(
    core: CoreState<'_>,
    base_revision: u64,
) -> Result<CloudSyncStatus, BackendError> {
    core.cloud_sync_delete(base_revision).await
}
